tool/tsh/common/hardware_key_test.go

```diff
@@ -0,0 +1,173 @@
+			cfg.Auth.NetworkingConfig.SetProxyListenerMode(types.ProxyListenerMode_Multiplex)
+		})
+	})
+	authServer := testServer.GetAuthServer()
+
+	proxyAddr, err := testServer.ProxyWebAddr()
+	require.NoError(t, err)
+
+	// mock SSO login and count the number of login attempts.
+	var loginCount int
+	setMockSSOLogin := func(cf *CLIConf) error {
+		loginCount = 0
+		mockSSOLogin := mockSSOLogin(t, authServer, alice)
+		cf.MockSSOLogin = func(ctx context.Context, connectorID string, priv *keys.PrivateKey, protocol string) (*auth.SSHLoginResponse, error) {
+			fmt.Println(priv.GetPrivateKeyPolicy())
+			loginCount++
+			return mockSSOLogin(ctx, connectorID, priv, protocol)
+		}
+		cf.AuthConnector = connector.GetName()
+		return nil
+	}
+
+	t.Run("cap", func(t *testing.T) {
+		setRequireMFAType := func(t *testing.T, requireMFAType types.RequireMFAType) {
+			// Set require MFA type in the cluster auth preference.
+			authPref, err := authServer.UpsertAuthPreference(ctx, &types.AuthPreferenceV2{
+				Spec: types.AuthPreferenceSpecV2{
+					RequireMFAType: requireMFAType,
+				},
+			})
+			require.NoError(t, err)
+
+			// Set MockAttestationData to attest the expected key policy. This will
+			// apply to the next (re)login.
+			testModules.MockAttestationData = &keys.AttestationData{
+				PrivateKeyPolicy: authPref.GetPrivateKeyPolicy(),
+			}
+		}
+
+		t.Cleanup(func() {
+			setRequireMFAType(t, types.RequireMFAType_OFF)
+		})
+
+		// login should use the private key policy reported by the proxy without
+		// needing to retry hardware key login.
+		setRequireMFAType(t, types.RequireMFAType_HARDWARE_KEY_TOUCH)
+		tmpHomePath := t.TempDir()
+		err = Run(context.Background(), []string{
+			"login",
+			"--debug",
+			"--insecure",
+			"--proxy", proxyAddr.String(),
+		}, setHomePath(tmpHomePath), setMockSSOLogin)
+		require.NoError(t, err)
+		require.Equal(t, 1, loginCount)
+
+		// Upgrading the auth preference requireMFAType should trigger relogin
+		// on the next command run.
+		setRequireMFAType(t, types.RequireMFAType_HARDWARE_KEY_TOUCH_AND_PIN)
+		err = Run(context.Background(), []string{
+			"ls",
+			"--debug",
+			"--insecure",
+			"--proxy", proxyAddr.String(),
+		}, setHomePath(tmpHomePath), setMockSSOLogin)
+		require.NoError(t, err)
+		require.Equal(t, 1, loginCount)
+	})
+
+	t.Run("role", func(t *testing.T) {
+		setRequireMFAType := func(t *testing.T, requireMFAType types.RequireMFAType) {
+			// Set require MFA type in the user's role.
+			aliceRole.SetOptions(types.RoleOptions{
+				RequireMFAType: requireMFAType,
+			})
+			_, err = authServer.UpsertRole(ctx, aliceRole)
+			require.NoError(t, err)
+
+			// Set MockAttestationData to attest the expected key policy. This will
+			// apply to the next (re)login.
+			testModules.MockAttestationData = &keys.AttestationData{
+				PrivateKeyPolicy: aliceRole.GetPrivateKeyPolicy(),
+			}
+		}
+
+		t.Cleanup(func() {
+			setRequireMFAType(t, types.RequireMFAType_OFF)
```

Comment:
Is OFF the default? Should the cleanup remove the option entirely instead?

Reply:
Yes, `OFF` is 0. It should have been named `UNSPECIFIED` but this was oh so long ago. So this is equivalent to removing the option.

Reply:
Gotcha, thanks.